internal/rhp/v4/transport.go

```diff
@@ -0,0 +1,102 @@
+package rhp
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"go.sia.tech/core/types"
+	rhp "go.sia.tech/coreutils/rhp/v4"
+)
+
+type transportPool struct {
+	dialer Dialer
+
+	mu   sync.Mutex
+	pool map[string]*transport
+}
+
+func newTransportPool(dialer Dialer) *transportPool {
+	return &transportPool{
+		dialer: dialer,
+		pool:   make(map[string]*transport),
+	}
+}
+
+func (p *transportPool) withTransport(ctx context.Context, hk types.PublicKey, addr string, fn func(rhp.TransportClient) error) (err error) {
+	// fetch or create transport
+	p.mu.Lock()
+	t, found := p.pool[addr]
+	if !found {
+		t = &transport{
+			dialer: p.dialer,
+			hk:     hk,
+			addr:   addr,
+		}
+		p.pool[addr] = t
+	}
+	t.refCount++
+	p.mu.Unlock()
+
+	// execute function
+	err = func() (err error) {
+		defer func() {
+			if r := recover(); r != nil {
+				err = fmt.Errorf("panic (withTransport): %v", r)
+			}
+		}()
+		client, err := t.Dial(ctx)
+		if err != nil {
+			return err
+		}
+		return fn(client)
+	}()
+
+	// Decrement refcounter again and clean up pool.
+	p.mu.Lock()
+	t.refCount--
+	if t.refCount == 0 {
+		// Cleanup
+		if t.t != nil {
+			_ = t.t.Close()
+			t.t = nil
+		}
+		delete(p.pool, addr)
+	}
+	p.mu.Unlock()
+	return err
+}
+
+type transport struct {
```

Comment:
I understand we just copied but I'm not sure why a `transport` isn't a ref count and a client.

```go
type (
	transportPool struct {
		dialer Dialer
		logger *zap.SugaredLogger

		mu   sync.Mutex
		pool map[string]*transport
	}

	transport struct {
		refCount uint64 // locked by pool
		tc       rhp4.TransportClient
	}
)
```

We always immediately `Dial` anyway.

Reply:
Because I didn't call `Dial` immediately before and did that in the client method instead but you are right. It can be simplified a bit.